# DS 3000 - Practicum 2

**Student Name**: Hwijoon Lee, Maya Karintholil, Sanya Mittal, and Shreyashi Kalakuntla

**Date**: 2025-02-21


### Submission Instructions
Submit this `ipynb` file to canvas.

The `ipynb` format stores outputs from the last time you ran the notebook.  (When you open a notebook it has the figures and outputs of the last time you ran it too).  To ensure that your submitted `ipynb` file represents your latest code, make sure to give a fresh run `Kernel > Restart & Run All` just before uploading the `ipynb` file to Canvas.

### Academic Integrity

**Writing your homework is an individual effort.**  You may discuss general python problems with other students but under no circumstances should you observe another student's code which was written for this assignment, from this year or past years.  Pop into office hours or DM us in MS Teams if you have a specific question about your work or if you would like another pair of eyes or talk through your code.

Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment. You do not need to cite the official python documentation.

**Documentation / style counts for credit**  Please refer to the Pep-8 style, to improve the readability and consistency of your Python code. For more information, read the following article [How to Write Beautiful Python Code With PEP 8](https://realpython.com/python-pep8/) or ask your TA's for tips.

**NOTE:<span style='color:red'> Write python expressions to answer ALL questions below and ensure that you use the `print()` function to display the output.</span>** Each question should be answered in a new code cell. For example, your solution for question 1.1 should be in a different code cell from your solution for question 1.2.

## Question 1: 

In your own words, formulate the Data Science problem that you were given.

In [1]:
#Write your python code to answer question 1. You can add new cells below as needed.

## Question 2: 
Load the data using pandas and inspect it.

Perform the initial inspection of the data, its shape, types, etc.
Evaluate the dataset and perform at least three type of data preparation and justify the approach that is taken to prepare the data for analysis. Data prep can include, but is not limited to: handling missing values, data types, duplicates, etc. You will need to ensure that your data preparation addressed issues in at least 7 fields in the data.
Prepare meaningful* summary statistics for 3 continuous variables and 3 categorical variables.
Note: meaningful summary statistics explains the statistical summary of relevant fields in a coherent manner.

In [2]:
import pandas as pd
import numpy as np

df = pd.read_csv('Bird_Strikes_1990_2023.csv')

/var/folders/dw/7lckm6l56sn70yq_dk59qxbr0000gn/T/ipykernel_4521/2048397694.py:4: DtypeWarning: Columns (18,21) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('Bird_Strikes_1990_2023.csv')


In [3]:
print("\nMissing values in each column:")
print(df.isnull().sum())


Missing values in each column:
INDEX_NR               0
INCIDENT_DATE          0
INCIDENT_MONTH         0
INCIDENT_YEAR          0
TIME              125177
                   ...  
REPORTED_TITLE         0
SOURCE                 0
PERSON             21134
LUPDATE                0
TRANSFER               0
Length: 100, dtype: int64


1. Convert INCIDENT_DATE to datetime type

In [4]:
df['INCIDENT_DATE'] = pd.to_datetime(df['INCIDENT_DATE'])

2. Handle missing values filling in Unknown for categorical variables and median for numeric variables
- We use median instead of mean because Median is less sensitive to extreme outliers

In [5]:
# Fill missing numeric values
df['LATITUDE'] = df['LATITUDE'].fillna(df['LATITUDE'].median())
df['LONGITUDE'] = df['LONGITUDE'].fillna(df['LONGITUDE'].median())
df['DISTANCE'] = df['DISTANCE'].fillna(df['DISTANCE'].median())

# Fill missing categorical values
df['STATE'] = df['STATE'].fillna('Unknown')
df['FAAREGION'] = df['FAAREGION'].fillna('Unknown')
df['TIME_OF_DAY'] = df['TIME_OF_DAY'].fillna('Unknown')

3. Removing completely empty columns

In [6]:
print("Before removal:")
print("Number of columns:", len(df.columns))

print("\nNon-null value counts for specific columns:")
print("LUPDATE:", df['LUPDATE'].count())
print("TRANSFER:", df['TRANSFER'].count())

print("\nTotal number of rows:", len(df))

# Remove unnecessary columns and duplicates
columns_to_drop = ['LUPDATE', 'TRANSFER']
df = df.drop(columns=columns_to_drop)
df = df.drop_duplicates()

print("\nAfter removal:")
print("Number of columns:", len(df.columns))


Before removal:
Number of columns: 100

Non-null value counts for specific columns:
LUPDATE: 288810
TRANSFER: 288810

Total number of rows: 288810

After removal:
Number of columns: 98


4. Summary Statistics for Continuous Variables:

In [7]:
continuous_vars = ['LATITUDE', 'LONGITUDE', 'DISTANCE']
print("\nSummary Statistics for Continuous Variables:")
print(df[continuous_vars].describe())


Summary Statistics for Continuous Variables:
           LATITUDE      LONGITUDE       DISTANCE
count  2.888100e+05  288810.000000  288810.000000
mean   1.813044e+02     -90.514448       0.530834
std    7.658451e+04     306.472863       2.890473
min   -3.767333e+01    -177.381000       0.000000
25%    3.367566e+01     -97.362440       0.000000
50%    3.880581e+01     -87.904460       0.000000
75%    4.077724e+01     -81.316030       0.000000
max    4.115443e+07  164140.000000      99.000000


Continuous Variables:
- The geographic data shows LATITUDE and LONGITUDE contain outliers beyond valid ranges, requiring cleanup before spatial analysis. 
- DISTANCE measurements reveal most bird strikes (over 50%) occur at 0 distance from airports, with values ranging up to 99 units.

#### Further clearning for Latitude and Longitude
- Latidue contains outliers beyond valid ranges, requiring cleanup before spatial analysis.

In [9]:
df['LATITUDE'] = df['LATITUDE'].apply(
    lambda x: x if (-90 <= x <= 90) else df['LATITUDE'].median()
)
df['LONGITUDE'] = df['LONGITUDE'].apply(
    lambda x: x if (-180 <= x <= 180) else df['LONGITUDE'].median()
)

# Display cleaned summary statistics
print("Cleaned Geographic Coordinates Summary:")
print(df[['LATITUDE', 'LONGITUDE']].describe())

Cleaned Geographic Coordinates Summary:
            LATITUDE      LONGITUDE
count  288810.000000  288810.000000
mean       37.108369     -91.083085
std         6.419505      23.152147
min       -37.673333    -177.381000
25%        33.675660     -97.362440
50%        38.805810     -87.904460
75%        40.777240     -81.316030
max        71.285450     178.559228


5. Summary Statistics for Categorical Variables:

In [ ]:
categorical_vars = ['STATE', 'FAAREGION', 'TIME_OF_DAY']
print("\nSummary Statistics for Categorical Variables:")
for var in categorical_vars:
    print(f"\nValue counts for {var}:")
    print(df[var].value_counts().head())
    print(f"Number of unique values: {df[var].nunique()}")


Summary Statistics for Categorical Variables:

Value counts for STATE:
STATE
Unknown    35501
TX         24794
FL         20616
CA         20431
NY         14371
Name: count, dtype: int64
Number of unique values: 68

Value counts for FAAREGION:
FAAREGION
ASO        53932
AEA        40845
AGL        39773
Unknown    35501
ASW        33354
Name: count, dtype: int64
Number of unique values: 11

Value counts for TIME_OF_DAY:
TIME_OF_DAY
Unknown    122416
Day        102548
Night       50446
Dusk         7428
Dawn         5972
Name: count, dtype: int64
Number of unique values: 5


Categorical Variables: 
- Texas, Florida, and California report the highest number of incidents among states, with 68 unique states/territories recorded. 
- The FAA regions show ASO (Southern) leading with 53,932 incidents, followed by AEA (Eastern) with 40,845. 
- TIME_OF_DAY analysis reveals most incidents occur during Unknown and Day periods, with 5 distinct time categories identified.

## Question 3: 
Perform an in-depth analysis by creating visualizations to understand the distribution and relationships within the data, while ensuring that your analysis and explanation of the results tell a story. It's important to select appropriate visualizations based on the type of data. Explain all results.

Analyze the proportion of incidents that occur each year.
Evaluate the distribution of bird strikes in each state and display the top 20 states.
What is the most common bird species that are involved in bird strikes.
Perform a comparative analysis of the annual aircraft damage for each phase of flight. Evaluate the results over the last 10 years to determine what patterns exist. Note: this requires a year over year analysis.
Analyze the frequency of bird strike incidents that occur on a monthly basis over the last 10 years. Analyze the results to determine when these incidents are more prevalent.  This question requires you to drill deeper into the analysis to determine if the distribution has changed over the last 10 years. Note: this requires a year over year analysis.
Perform ONE (1) additional analysis on the distribution. 

Ensure that you either : 1. perform a comparative analysis by evaluating 3 or more groups within the data or 2. perform a temporal analysis of the patterns over 10 or more years (this requires a year over year analysis of the patterns). Do not repeat any of the previous analysis.

## Question 4: 
Perform pairwise analysis of select features and evaluate the significance of the pattern or trend. A suitable value for alpha is 5%. Explain all results.

Create a scatterplot that shows the relationship between aircraft height and speed. Evaluate the correlation, the strength and the significance of the results.
Visualize the distribution of the aircraft speed during: 1. the approach phase of flight and 2. the landing roll phase of flight. Perform a 2 sample t-test and evaluate if there is a statistical difference between the speed during these two flight phases. Tip: if the data is skewed, you will need to address this prior to the statistical analysis.
Create a visualization of the aircraft damage grouped by phase of flight.
Evaluate if the results are statistically significant. Ensure that you use the appropriate test.
Perform ONE (1) additional statistical test.
Explain what you are testing and the reason this information is useful.
Visualize the data, state the hypothesis and explain if it is statistically significant.

## Question 5: 
Explore a simple linear relationship* within the data and create a model to predict the occurrence of bird strikes in a given FAA region. Ensure that you explain all results.

Extract all data for the AWP FAA Region and use a line chart to visualize the frequency of incidents between 1990 and 2015 inclusive.
Using the extracted data above, create a simple linear regression model and predicts the incidents for 2016 and 2017.
Display the r-squared and what this indicates about your model's ability to fit the data.
Visualize the dataset from (1) and include the predicted results for 2016 and 2017.
Compare and contrast the predictions for 2016 and 2017 with the actual values.
Perform predictions for 2018 through to 2020. Explain what transpired when your model was used to predict this time period. Comment on the reliability of your model.
*Tip: It is important to remember that you are creating a simple linear regression model, which is limited in features and does not include information about all factors that influence the frequency of bird strikes. Keep this in mind while framing your response and explanations

## Question 6: 
Summarize the overall analysis for questions 1 through 5 and share THREE (3) key findings.

## Helpful resources 
Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment inside the code cell, or you can list them below. 

You do not need to cite the official python documentation.
